### PR TITLE
Added StackedMap data structure

### PR DESCRIPTION
**Pull request by lwoodson**
A stack of Map's that itself implements a Map interface.  Its
function as a Map is of a flattened version of itself, containing
all keys/values from all maps in the stack.  Where key collisions
occur between maps in the stack, the value closer to the top of
the stack wins.

This data structure is a requisite for [GumShoe's context behavior](https://github.com/devquixote/gumshoe#contexts).

### CODE REVIEW

**Comment by lwoodson** [comment]
Merging now, will address code review feedback in a subsequent PR.

File: java/src/main/java/com/bazaarvoice/gumshoe/StackedMap.java [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+import java.util.EmptyStackException;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.Stack;
+
+/**
+ * A stack of Map's that can be flattened into a single Map.  Key
```
Comment (aoudenne) [inline java/src/main/java/com/bazaarvoice/gumshoe/StackedMap.java:14]: Map's -> Maps

File: java/src/main/java/com/bazaarvoice/gumshoe/StackedMap.java [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+     * values override older values.
+     *
+     * @return
+     */
+    public Map<K, V> flatten() {
+        Map<K, V> result = new HashMap<K, V>();
+        Iterator<Map<K, V>> iterator = stack.iterator();
+
+        // Java's stack is a subclass of Vector, so iteration order
+        // is FIFO, not LIFO
```
Comment (aoudenne) [inline java/src/main/java/com/bazaarvoice/gumshoe/StackedMap.java:104]: That is bizarre to me.  Good to know!

**Comment by aoudenne** [comment]
LGTM!  I did have to spend a couple of minutes figuring out the difference between push and put.  I understand now, but it might be worth a comment on the put method-- initially I thought it was going to make a map out of the key and value it accepts, but then I realized it was adding them to the map on the top of the stack (and if that's not what it is doing, then you _definitely_ need a comment!).

File: java/src/main/java/com/bazaarvoice/gumshoe/StackedMap.java (hunk 2 above)
Comment (lwoodson) [inline java/src/main/java/com/bazaarvoice/gumshoe/StackedMap.java:104]: Yeah, surprised me as well.

**Comment by lwoodson** [comment]
That's correct as to what its doing, @aoudenne.  I'll add a comment.

**Pull request merged**